### PR TITLE
feat: introduces conditional control plane installation

### PR DESCRIPTION
**Pull request by bartoszmajsak**
## Overview

When using KfDef such as:

```
apiVersion: kfdef.apps.kubeflow.org/v1
kind: KfDef
metadata:
  name: odh-mesh
spec:
  plugins:
    - kind: KfOssmPlugin
      spec:
        mesh:
          name: minimal
          namespace: istio-system
          mode: Managed
          certificate:
            generate: true
        auth:
          name: authorino
          namespace: auth-provider
          authorino:
            label: authorino/topic=odh
  applications:
    # your components
  repos:
    - name: manifests
      uri: https://github.com/maistra/odh-manifests/tarball/v0.0.6
  version: v0.0.6
```

the operator will install a minimal version of the control plane.

## Changes

- new feature of conditional installation of control plane based on `Mesh.Mode` value
- reworked plugin integration in the `KfApp` call chain (see docs around `Apply()` and `Delete()`)
- fixed issue of leaving KfApp in an odd state when errors on Apply occurs (raised as separated PR too https://github.com/opendatahub-io/opendatahub-operator/pull/501)
- reworks SMCP readiness check - it now checks not status, but a list of components in defined states
- simplifies CRD existence check call

### CODE REVIEW

File: pkg/kfapp/ossm/feature/builder.go [hunk 1]
```diff
@@ -53,6 +54,10 @@ func (fb *featureBuilder) UsingConfig(config *rest.Config) *featureBuilder {
 			return errors.WithStack(err)
 		}
 
+		if err := apiextv1.AddToScheme(f.client.Scheme()); err != nil {
```
Comment (bartoszmajsak) [inline pkg/kfapp/ossm/feature/builder.go:57]: We need this as we now query for CRD resource

File: pkg/kfapp/ossm/feature/verification.go [hunk 2]
```diff
@@ -2,69 +2,77 @@ package feature
...
-func checkSMCPStatus(dynamicClient dynamic.Interface, name, namespace string) (string, error) {
-	gvr := schema.GroupVersionResource{
-		Group:    "maistra.io",
-		Version:  "v1",
-		Resource: "servicemeshcontrolplanes",
-	}
+const (
+	interval = 1 * time.Second
+	duration = 5 * time.Minute
+)
```
Comment (cam-garrison) [inline pkg/kfapp/ossm/feature/verification.go]: Checking every second for 5 minutes seems like overkill to me.

File: pkg/kfapp/ossm/templates/control-plane/control-plane-managed.tmpl [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+metadata:
+  name: {{ .Mesh.Name }}
+  namespace: {{ .Mesh.Namespace }}
+spec:
+  version: v2.4
+  tracing:
+    type: None
+  addons:
+    prometheus:
+      enabled: true
```
Comment (cam-garrison) [inline pkg/kfapp/ossm/templates/control-plane/control-plane-managed.tmpl]: Curious as to why Prometheus is enabled in this SMCP - in our sample in `enabling-ossm.md` we have it set to [false](https://github.com/maistra/odh-manifests/blob/682f627a4047f9cbb036d2d9d70d17b63b9a9e60/enabling-ossm.md?plain=1#L91).

**Review by cam-garrison — CHANGES_REQUESTED** [review]
Requesting changes (or explanations) for two things, otherwise LGTM!

File: pkg/kfapp/ossm/feature/verification.go (hunk 2 above)
Comment (bartoszmajsak) [reply pkg/kfapp/ossm/feature/verification.go]: What should we bump here?

File: pkg/kfapp/ossm/templates/control-plane/control-plane-managed.tmpl (hunk 3 above)
Comment (bartoszmajsak) [reply pkg/kfapp/ossm/templates/control-plane/control-plane-managed.tmpl]: There are no requirements as of now what and if self install is needed. I thought this is helpful as we have metrics from istio collected

File: pkg/kfapp/ossm/feature/verification.go (hunk 2 above)
Comment (cam-garrison) [reply pkg/kfapp/ossm/feature/verification.go]: I was thinking that checking every 1 second was excessive - maybe every 5/10 seconds?

File: pkg/kfapp/ossm/templates/control-plane/control-plane-managed.tmpl (hunk 3 above)
Comment (cam-garrison) [reply pkg/kfapp/ossm/templates/control-plane/control-plane-managed.tmpl]: Gotcha, that makes sense to me :+1:

File: pkg/kfapp/ossm/feature/verification.go [hunk 4]
```diff
@@ -2,69 +2,77 @@ package feature
...
 }
 
-func checkSMCPStatus(dynamicClient dynamic.Interface, name, namespace string) (string, error) {
-	gvr := schema.GroupVersionResource{
-		Group:    "maistra.io",
-		Version:  "v1",
-		Resource: "servicemeshcontrolplanes",
-	}
+const (
+	interval = 1 * time.Second
```
Comment (bartoszmajsak) [inline pkg/kfapp/ossm/feature/verification.go]:
```suggestion
	interval = 5 * time.Second
```

**Comment by cam-garrison** [comment]
Note from testing. If the `mesh.mode` field is not set, eg:

```
mesh:
          name: basic
          namespace: istio-system
          certificate:
            generate: true
```

The operator enters a CrashLoopBackOff state, with the error 
```
2023-08-31T20:04:22.269Z INFO KfOssmPlugin Initializing {"plugin": "KfOssmPlugin"}
panic: reflect.Set: value of type string is not assignable to type ossmplugin.InstallationMode
goroutine 813 [running]:
reflect.Value.assignTo({0x1e65ba0?, 0xc0038441c0?, 0x7f992d423f18?}, {0x22e0a70, 0xb}, 0x1e5fee0, 0x0)
/usr/lib/golang/src/reflect/value.go:3064 +0x2ac
reflect.Value.Set({0x1e5fee0?, 0xc005a70180?, 0x22d975a?}, {0x1e65ba0?, 0xc0038441c0?, 0x1e5fee0?})
/usr/lib/golang/src/reflect/value.go:2090 +0xeb
github.com/opendatahub-io/opendatahub-operator/pkg/kfconfig/ossmplugin.setDefaults({0x1f9bea0?, 0xc005a70160?})
/workspace/pkg/kfconfig/ossmplugin/types.go:89 +0x2fe
github.com/opendatahub-io/opendatahub-operator/pkg/kfconfig/ossmplugin.setDefaults({0x202bea0?, 0xc005a70160?})
/workspace/pkg/kfconfig/ossmplugin/types.go:93 +0x33b
github.com/opendatahub-io/opendatahub-operator/pkg/kfconfig/ossmplugin.(*OssmPluginSpec).SetDefaults(...)
..................
```

To get out of it, have to delete KfDef and remove finalizer manually.

**Comment by bartoszmajsak** [comment]
> The operator enters a CrashLoopBackOff state, with the error

Thanks for spotting it! It's now fixed in ca7e9d8

in YAML: `mode` -> `installationMode`

values: 

  * `pre-installed`
  * `minimal`

prometheus is gone

**Review by cam-garrison — APPROVED** [review]
tested some more, lgtm

**Pull request merged**